### PR TITLE
[Games] Improvements for libretro shader presets support

### PR DESCRIPTION
**Pull request by KOPRajs**
## Description

This PR adds support for the video shader pass aliases and fixes wrong _FrameDirection_ GL uniform type.

## Motivation and context

There was a hope that adding the shader pass alias support will make the _CRT-Royale_ shader work, but unfortunately, it turned out that it is not enough (it still renders black screen). At least we've got a few steps closer to the full RetroArch shader compatibility and I've found and fixed several other minor issues along the way.

Most notable is the fix for the _FrameDirection_ GL uniform type. It was copied from the Windows version, but the Windows version is using _CG_ shaders and the specification for the _CG_ shaders differs from the _GLSL_ (_float_ vs _int_). The _VHS_ shader preset was triggering a GL error because of the wrong type. There are no more GL errors after the fix and the _VHS_ preset now shows the blinking "Play" OSD like in RetroArch.

EDIT: There are two more items remaining on my _TODO_ list regarding the video shader preset support:

1. The LUTs are currently being loaded for each pass separately, but according to the specification they should be common for all passes in the preset.

2. We should move all calls to `glGetUniformLocation()` from the `CShaderGL::SetShaderParameters()` (which is called every frame) to the `CShaderGL::GetUniformLocs()`. We need to cache the values in some kind of array and reuse them when doing the rendering.

## How has this been tested?

Linux / Wayland / OpenGL and GLES

## What is the effect on users?

Fully working _VHS_ shader preset (and possibly other presets using the _FrameDirection_ uniform) and added support for shader preset pass aliases.

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/cores/RetroPlayer/shaders/gl/ShaderGL.cpp [hunk 1]
```diff
@@ -402,29 +405,48 @@ void CShaderGL::SetShaderParameters(CShaderTextureGL& sourceTexture)
...
+      paramLoc = glGetUniformLocation(m_shaderProgram, (paramPassPrev + "Texture").c_str());
+      glUniform1i(paramLoc, textureUnit);
+
+      if (i > 0)
+      {
+        std::string paramAlias = m_passesUniformFrameInputs[i - 1].alias;
+        if (paramAlias != "")
+        {
+          paramLoc = glGetUniformLocation(m_shaderProgram, (paramAlias + "InputSize").c_str());
+          glUniform2f(paramLoc, m_passesUniformFrameInputs[i].input_size.x,
```
[the remark this answers is not in the document]
Comment (KOPRajs) [reply xbmc/cores/RetroPlayer/shaders/gl/ShaderGL.cpp:437]: This is intentional. The struct holds information about the source texture of the shader pass, but the aliased name should point to the target texture (which is the source texture of the next shader pass). Because we are looping through the texture indexes, we need to use the alias from previous pass.

File: xbmc/cores/RetroPlayer/shaders/gles/ShaderGLES.cpp [hunk 2]
```diff
@@ -387,29 +390,48 @@ void CShaderGLES::SetShaderParameters(CShaderTextureGLES& sourceTexture)
...
+      paramLoc = glGetUniformLocation(m_shaderProgram, (paramPassPrev + "Texture").c_str());
+      glUniform1i(paramLoc, textureUnit);
+
+      if (i > 0)
+      {
+        std::string paramAlias = m_passesUniformFrameInputs[i - 1].alias;
+        if (paramAlias != "")
+        {
+          paramLoc = glGetUniformLocation(m_shaderProgram, (paramAlias + "InputSize").c_str());
+          glUniform2f(paramLoc, m_passesUniformFrameInputs[i].input_size.x,
```
[the remark this answers is not in the document]
Comment (KOPRajs) [reply xbmc/cores/RetroPlayer/shaders/gles/ShaderGLES.cpp:422]: The same as above for GLES.

File: xbmc/cores/RetroPlayer/shaders/gles/ShaderGLES.cpp [hunk 3]
```diff
@@ -387,29 +390,48 @@ void CShaderGLES::SetShaderParameters(CShaderTextureGLES& sourceTexture)
...
+      paramLoc = glGetUniformLocation(m_shaderProgram, (paramPassPrev + "TextureSize").c_str());
+      glUniform2f(paramLoc, m_passesUniformFrameInputs[i].texture_size.x,
+                  m_passesUniformFrameInputs[i].texture_size.y);
+      paramLoc = glGetUniformLocation(m_shaderProgram, (paramPassPrev + "Texture").c_str());
+      glUniform1i(paramLoc, textureUnit);
+
+      if (i > 0)
+      {
+        std::string paramAlias = m_passesUniformFrameInputs[i - 1].alias;
+        if (paramAlias != "")
```
[the remark this answers is not in the document]
Comment (KOPRajs) [reply xbmc/cores/RetroPlayer/shaders/gles/ShaderGLES.cpp]: The same as above for GLES.

[the remark this answers is not in the document]
Comment (garbear) [reply xbmc/cores/RetroPlayer/shaders/gles/ShaderGLES.cpp]: What do you refer to? We prefer `.empty()` over `!= ""`.

File: xbmc/cores/RetroPlayer/shaders/gl/ShaderGL.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (garbear) [reply xbmc/cores/RetroPlayer/shaders/gl/ShaderGL.cpp:437]: Is this deserving of a comment in the code?

**Comment by garbear** [comment]
> the VHS preset now shows the blinking "Play" OSD like in RetroArch.

I can confirm that this functionality now works on my macOS build.

Overall the fixes look great. The first pass alias commit looks good, but I'm not familiar enough with GL to pass judgement on the second. Claude only found a suggestion for `.empty()`, and maybe we add a comment about the index issue.

**Comment by garbear** [comment]
@claude please review

File: xbmc/cores/RetroPlayer/shaders/gl/ShaderGL.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (KOPRajs) [reply xbmc/cores/RetroPlayer/shaders/gl/ShaderGL.cpp:437]: I think a single comment on this line wouldn't be very useful. We would have to document the logic as a whole somewhere. Unfortunately, even the libretro's documentation in this area is weak. I've had to study the logic from the RetroArch sources, which is full of these _+1_ and _-1_ everywhere.

There is actually one more important thing to say. After implementing this I've been testing the _CRT-Royale_ and have found out, that it still doesn't work. I tried to find another simpler preset with alias to test, but it turned out, there is none. The alias feature was introduced to RA around the same time as they started the switch to Slang shaders. It was implemented for both GLSL and Slang, but all the GLSL shaders in the libretro's default library are kept compatible with the pre-alias code. Turned out that even though there are aliases specified in the preset files for many presets, none of the shaders actually depends on them (including the _CRT-Royale_). While I'm pretty sure that the code is correct (and might be useful for user-downloaded presets), I've got not a single shader preset to actually verify it.

File: xbmc/cores/RetroPlayer/shaders/gles/ShaderGLES.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (KOPRajs) [reply xbmc/cores/RetroPlayer/shaders/gles/ShaderGLES.cpp]: The same change is suggested for both GL and GLES.

Anyway, I've used `""`, because it is set explicitly to `""` here:
https://github.com/xbmc/xbmc/blob/515f20eccfa7c59b92c0a1abaf1bf0c241e71374/xbmc/addons/ShaderPreset.cpp#L219

But if you want me to change it to `.empty()`, let me know.

File: xbmc/cores/RetroPlayer/shaders/gl/ShaderGL.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (garbear) [reply xbmc/cores/RetroPlayer/shaders/gl/ShaderGL.cpp:437]: Let's ship the alias code, even if a bug sneaks up, it puts us closer to supporting user-downloaded presets with aliases.

File: xbmc/cores/RetroPlayer/shaders/gles/ShaderGLES.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (garbear) [reply xbmc/cores/RetroPlayer/shaders/gles/ShaderGLES.cpp]: Yes, please use `.empty()`. `""` is on the "C" side, and assigned to the "C++" side, to get it into C++ style as quick as possible, so we want to treat it in C++ style everywhere else that we can.

**Review by garbear — APPROVED** [review]
Fix is confirmed and the alias work is a good progression of the code. Once the `.empty()`s are replaced this can go in.

**Comment by KOPRajs** [comment]
> Fix is confirmed and the alias work is a good progression of the code. Once the `.empty()`s are replaced this can go in.

Done.
